vizro-core/src/vizro/managers/_model_manager.py

```diff
@@ -80,39 +93,69 @@ def _get_models(
 
         if model_type is FIGURE_MODELS:
             model_type = (vm.Graph, vm.AgGrid, vm.Table, vm.Figure)  # type: ignore[assignment]
-        models = self.__get_model_children(root_model) if root_model is not None else self.__models.values()
 
-        # Convert to list to avoid changing size when looping through at runtime.
-        for model in list(models):
+        if root_model is not None:
+            root_node = self.__dashboard_tree.find(data_id=root_model.id)
+            nodes = root_node.iterator(method=IterMethod.PRE_ORDER) if root_node else []
+        else:
+            nodes = self.__dashboard_tree.iterator(method=IterMethod.PRE_ORDER)
+
+        for node in nodes:
+            model = node.data
             if model_type is None or isinstance(model, model_type):
-                yield model  # type: ignore[misc]
+                yield model
 
-    def __get_model_children(self, model: Model) -> Generator[Model, None, None]:
+    def __populate_tree(
+        self, model: Union[Model, Mapping[str, Model], Collection[Model]], parent=None, field_name=None
+    ):
         """Iterates through children of `model` with depth-first pre-order traversal."""
+        # AM rough notes: this is basically copied and pasted from __get_model_children and then modified to populate
+        # dashboard tree instead of yielding.
+        # MS: This misses items created in pre-build (like selectors) and "behind the scene" items like ActionsChain
```

Comment:
These are now added by `_from_dict_in_build`, right? Is there anything else still missing from `__dashboard_tree` after `pre_build` has completed?

Would be worth checking we haven't missed anything here by trying it on some complicated example dashboards.